File: tests/bench/error_paths/ErrorPaths.java
// Same program as error_paths.jai.
public class ErrorPaths {
    static long scale() {
        String l = System.getenv("BENCH_LEVEL");
        if ("easy".equals(l)) return 16;
        if ("medium".equals(l)) return 4;
        return 1;
    }

    static final long SCALE = scale();
    static final long CALM = Math.max(8_000_000L / SCALE, 1);
    static final long STORM = Math.max(160_000L / SCALE, 1);

    static class BadValue extends RuntimeException {
        BadValue(String msg) { super(msg); }
    }

    static long checked(long x) {
        if (x % 8 == 3) throw new BadValue("bad value " + x);
        return x + 1;
    }

    static long calm(long n) {
        long total = 0;
        long i = 0;
        while (i < n) {
            try {
                total += i % 13;
            } catch (BadValue e) {
                total -= 1;
            }
            i += 1;
        }
        return total;
    }

    static long storm(long n) {
        long ok = 0;
        long bad = 0;
        long i = 0;
        while (i < n) {
            try {
                ok += checked(i);
            } catch (BadValue e) {
                bad += 1;
            }
            i += 1;
        }
        return ok - bad;
    }

    public static void main(String[] args) {
        System.out.println(calm(CALM));
        System.out.println(storm(STORM));
    }
}

File: tests/bench/error_paths/error_paths.py
import os

_LEVEL = os.environ.get("BENCH_LEVEL", "hard")
SCALE = 16 if _LEVEL == "easy" else 4 if _LEVEL == "medium" else 1
CALM = max(8_000_000 // SCALE, 1)
STORM = max(160_000 // SCALE, 1)


def checked(x):
    if x % 8 == 3:
        raise ValueError(f"bad value {x}")
    return x + 1


def calm(n):
    total = 0
    i = 0
    while i < n:
        try:
            total += i % 13
        except ValueError:
            total -= 1
        i += 1
    return total


def storm(n):
    ok = 0
    bad = 0
    i = 0
    while i < n:
        try:
            ok += checked(i)
        except ValueError:
            bad += 1
        i += 1
    return ok - bad


print(calm(CALM))
print(storm(STORM))
